### DISCUSSION TITLE
SEnding MIDI message command

### DISCUSSION THREAD

**Discussion by Bipman1** [melanchall/drywetmidi#240] [Q&A] [answered]
Hi

Amazing library and much better than some out there!  I've got my program sending notes etc. to a soft synth and that is all working well but what I don't know how to do is send a command.  For example, say a knob on a soft synth hosted in Ableton live is mapped to a specific MIDI number, how would I send a command to alter the knob?  I hope that makes sense and if someone could point me in the right direction I'd be most happy!

Bipman

**Comment by melanchall** [comment, marked as answer]
Hi,

> mapped to a specific MIDI number

Can you please clarify what do you mean? Are you talking about MIDI controller numbers (MIDI CC)? If yes, just try to send `ControlChangeEvent` like this:

```csharp
_outputDevice.SendEvent(new ControlChangeEvent(
    (SevenBitNumber)controllerNumber,
    (SevenBitNumber)controllerValue) { Channel = (FourBitNumber)channel });
```

Set `controllerNumber`, `controllerValue` and `channel` to required values. Please note that channel in DryWetMIDI is a number **from 0 to 15**, not from 1 to 16.

Thanks,
Max

**Comment by Bipman1** [comment]
Hi

I'll give it a go thanks.

Bipman

**Comment by Bipman1** [comment]
Hmmm, not working.  The item I am trying to change is a MIDI CC item if that makes a difference.  For instance CC 77 is vibrato depth and 93 is chorus depth.  Also, in your code above I couldn't get the channel to work as it says the constructor does not take 3 arguments.

Bipman

**Reply by melanchall** [reply]
Please provide your code. In my code example I didn't show you 3 parameters in constructor so I don't understand what are trying to do. Also please note that DryWetMIDI has complete documentation. For example, on `ControlChangeEvent` – https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Core.ControlChangeEvent.html.

**Comment by Bipman1** [comment]
Hi,  The code is 

using System;
using Melanchall.DryWetMidi.Multimedia;
using Melanchall.DryWetMidi.Core;
using System.Threading;

// ...

using (var outputDevice = OutputDevice.GetByName("Gloving"))
{
    outputDevice.EventSent += OnEventSent;

    using (var inputDevice = InputDevice.GetByName("Gloving"))
    {
        Melanchall.DryWetMidi.Common.SevenBitNumber controllerNumber = (Melanchall.DryWetMidi.Common.SevenBitNumber)93;
        Melanchall.DryWetMidi.Common.SevenBitNumber controllerValue = (Melanchall.DryWetMidi.Common.SevenBitNumber)0;
        Melanchall.DryWetMidi.Common.SevenBitNumber channel = (Melanchall.DryWetMidi.Common.SevenBitNumber)1;
        float count = 0;

        inputDevice.EventReceived += OnEventReceived;
        inputDevice.StartEventsListening();

        outputDevice.SendEvent(new NoteOnEvent((Melanchall.DryWetMidi.Common.SevenBitNumber)60, (Melanchall.DryWetMidi.Common.SevenBitNumber)127));
        Thread.Sleep(1000);
        outputDevice.SendEvent(new NoteOffEvent());
        Thread.Sleep(500);
        outputDevice.SendEvent(new NoteOnEvent((Melanchall.DryWetMidi.Common.SevenBitNumber)60, (Melanchall.DryWetMidi.Common.SevenBitNumber)127));
        Thread.Sleep(1000);

        for (count = 0; count < 1; count += 0.1F)
        {

            outputDevice.SendEvent(new ControlChangeEvent(controllerNumber, controllerValue));
            controllerValue = (Melanchall.DryWetMidi.Common.SevenBitNumber)count;
            Thread.Sleep(200);
        }
        
        outputDevice.SendEvent(new NoteOffEvent());

    }
}



 void OnEventReceived(object sender, MidiEventReceivedEventArgs e)
{
    var midiDevice = (MidiDevice)sender;
    Console.WriteLine($"Event received from '{midiDevice.Name}' at {DateTime.Now}: {e.Event}");
}

 void OnEventSent(object sender, MidiEventSentEventArgs e)
{
    var midiDevice = (MidiDevice)sender;
    Console.WriteLine($"Event sent to '{midiDevice.Name}' at {DateTime.Now}: {e.Event}");
}//

It's going through loopMidi and I have the logs from there if it helps.  I'm going to switch on logging in Ableton too.

Bipman

[SPG-NEW12TH.LOG](https://github.com/melanchall/drywetmidi/files/10768208/SPG-NEW12TH.LOG)

**Comment by Bipman1** [comment]
Hi

The answer seems to be that although I am setting the controllerValue in the loop to the count variable, it remains at 0.  I'm obviously setting it wrong somehow.

Bipman

**Comment by Bipman1** [comment]
Hi

Got it working.  I was stupidly trying to send a float as the value on the synth can be 0.00 to 1.00.  I changed to int and all OK!  I'm used to using c++ (as an extension to C) as is probably evident in my code but c# has better libraries so am trying to switch from c++ to c# for when I don't need performance.  Great library by the way and thanks for the support.

Bipman

**Reply by melanchall** [reply]
Thank you :)
